pkg/operator/ceph/csi/spec.go

```diff
@@ -711,27 +759,36 @@ func (r *ReconcileCSI) stopDrivers(ver *version.Info) error {
 	return nil
 }
 
-func (r *ReconcileCSI) deleteCSIDriverResources(ver *version.Info, daemonset, deployment, service, driverName string) error {
-	csiDriverobj := v1CsiDriver{}
-	err := k8sutil.DeleteDaemonset(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace, daemonset)
-	if err != nil {
-		return errors.Wrapf(err, "failed to delete the %q", daemonset)
+func (r *ReconcileCSI) deleteCSIDriverResources(ver *version.Info, daemonset, deployment, service, driverName string, ownerInfo *k8sutil.OwnerInfo) error {
+	daemonOwnedByRook := r.daemonOwnedByRook(ownerInfo, daemonset)
+	if daemonOwnedByRook {
+		err := k8sutil.DeleteDaemonset(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace, daemonset)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete the %q", daemonset)
+		}
 	}
 
-	err = k8sutil.DeleteDeployment(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace, deployment)
-	if err != nil {
-		return errors.Wrapf(err, "failed to delete the %q", deployment)
+	provisionerOwnedByRook := r.provisionerOwnedByRook(ownerInfo, deployment)
+	if provisionerOwnedByRook {
+		err := k8sutil.DeleteDeployment(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace, deployment)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete the %q", deployment)
+		}
 	}
 
-	err = k8sutil.DeleteService(r.opManagerContext, r.context.Clientset, r.opConfig.OperatorNamespace, service)
-	if err != nil {
-		return errors.Wrapf(err, "failed to delete the %q", service)
-	}
+	if daemonOwnedByRook && provisionerOwnedByRook {
```

Comment:
The only assumption here is the service and csiDriverobj is created by rook only daemon and provisoner is created by rook.

Otherwise, we can also have an owner reference check with the service, but I am not sure if we can have such a thing for csiobject, and we need to add that field in the csi-driver repo.

Reply:
Yes, CSIDriver might not have owner ref as I don't think we have a clusterscope obj which can own CSIDriver obj.

If you really want tracking, you could add an annotation or label.

Reply:
Do think of a case where user deploying only the ceph csi object separately?

Reply:
users can create any secret or csi driver object with whatever name they want. lets not go with assumptions on what ceph csi does

Reply:
Okay yaa it totally make sense